# Data wrangling

## 1-Imports

In [271]:
#Import pandas
import pandas as pd
from library.sb_utils import save_file

## 2-Load the data


In [262]:
# Load The meal delivery company Data: Food Orders, Fulfilment center info, and meals info
orders = pd.read_csv('../Data/Food_orders_data.csv')
centers = pd.read_csv('../Data/fulfilment_center_info.csv')
meals= pd.read_csv('../Data/meal_info.csv')

## 3-Explore the data

In [263]:
#use the info method to see a summary of the data
orders.info()
centers.info()
meals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 456548 entries, 0 to 456547
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     456548 non-null  int64  
 1   week                   456548 non-null  int64  
 2   center_id              456548 non-null  int64  
 3   meal_id                456548 non-null  int64  
 4   checkout_price         456548 non-null  float64
 5   base_price             456548 non-null  float64
 6   emailer_for_promotion  456548 non-null  int64  
 7   homepage_featured      456548 non-null  int64  
 8   num_orders             456548 non-null  int64  
dtypes: float64(2), int64(7)
memory usage: 31.3 MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   center_id    77 non-null     int64  
 1   city_code    77 non-nul

In [264]:
#use the head method to see the first several rows of the data
orders.head()

id  week  center_id  meal_id  checkout_price  base_price  \
0  1379560     1         55     1885          136.83      152.29   
1  1466964     1         55     1993          136.83      135.83   
2  1346989     1         55     2539          134.86      135.86   
3  1338232     1         55     2139          339.50      437.53   
4  1448490     1         55     2631          243.50      242.50   

   emailer_for_promotion  homepage_featured  num_orders  
0                      0                  0         177  
1                      0                  0         270  
2                      0                  0         189  
3                      0                  0          54  
4                      0                  0          40

In [265]:
centers.head()

center_id  city_code  region_code center_type  op_area
0         11        679           56      TYPE_A      3.7
1         13        590           56      TYPE_B      6.7
2        124        590           56      TYPE_C      4.0
3         66        648           34      TYPE_A      4.1
4         94        632           34      TYPE_C      3.6

In [266]:
meals.head()

meal_id   category cuisine
0     1885  Beverages    Thai
1     1993  Beverages    Thai
2     2539  Beverages    Thai
3     1248  Beverages  Indian
4     2631  Beverages  Indian

In [267]:
#join all three tables
food_data=pd.merge(pd.merge(orders,centers, on='center_id'),meals, on='meal_id')
food_data.head()

id  week  center_id  meal_id  checkout_price  base_price  \
0  1379560     1         55     1885          136.83      152.29   
1  1018704     2         55     1885          135.83      152.29   
2  1196273     3         55     1885          132.92      133.92   
3  1116527     4         55     1885          135.86      134.86   
4  1343872     5         55     1885          146.50      147.50   

   emailer_for_promotion  homepage_featured  num_orders  city_code  \
0                      0                  0         177        647   
1                      0                  0         323        647   
2                      0                  0          96        647   
3                      0                  0         163        647   
4                      0                  0         215        647   

   region_code center_type  op_area   category cuisine  
0           56      TYPE_C      2.0  Beverages    Thai  
1           56      TYPE_C      2.0  Beverages    Thai  
2           56      TYPE_C      2.0  Beverages    Thai  
3           56      TYPE_C      2.0  Beverages    Thai  
4           56      TYPE_C      2.0  Beverages    Thai

In [268]:
#change 'category' column name 
food_data.rename(columns={'category':'meal_category'},inplace=True)

In [272]:
# Save the data 

datapath = '../data'
save_file(food_data, 'food_data_merged.csv', datapath)

Writing file.  "../data/food_data_merged.csv"


In [269]:
food_data.head()

id  week  center_id  meal_id  checkout_price  base_price  \
0  1379560     1         55     1885          136.83      152.29   
1  1018704     2         55     1885          135.83      152.29   
2  1196273     3         55     1885          132.92      133.92   
3  1116527     4         55     1885          135.86      134.86   
4  1343872     5         55     1885          146.50      147.50   

   emailer_for_promotion  homepage_featured  num_orders  city_code  \
0                      0                  0         177        647   
1                      0                  0         323        647   
2                      0                  0          96        647   
3                      0                  0         163        647   
4                      0                  0         215        647   

   region_code center_type  op_area meal_category cuisine  
0           56      TYPE_C      2.0     Beverages    Thai  
1           56      TYPE_C      2.0     Beverages    Thai  
2           56      TYPE_C      2.0     Beverages    Thai  
3           56      TYPE_C      2.0     Beverages    Thai  
4           56      TYPE_C      2.0     Beverages    Thai

In [199]:
#use the info method to see a summary of the data
food_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 456548 entries, 0 to 456547
Data columns (total 15 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     456548 non-null  int64  
 1   week                   456548 non-null  int64  
 2   center_id              456548 non-null  int64  
 3   meal_id                456548 non-null  int64  
 4   checkout_price         456548 non-null  float64
 5   base_price             456548 non-null  float64
 6   emailer_for_promotion  456548 non-null  int64  
 7   homepage_featured      456548 non-null  int64  
 8   num_orders             456548 non-null  int64  
 9   city_code              456548 non-null  int64  
 10  region_code            456548 non-null  int64  
 11  center_type            456548 non-null  object 
 12  op_area                456548 non-null  float64
 13  meal_category          456548 non-null  object 
 14  cuisine                456548 non-nu

The selected data types for the columns seem appropriate.

In [200]:
#Perform summary statistics on numeric features
food_data.describe()

id           week      center_id        meal_id  \
count  4.565480e+05  456548.000000  456548.000000  456548.000000   
mean   1.250096e+06      74.768771      82.105796    2024.337458   
std    1.443548e+05      41.524956      45.975046     547.420920   
min    1.000000e+06       1.000000      10.000000    1062.000000   
25%    1.124999e+06      39.000000      43.000000    1558.000000   
50%    1.250184e+06      76.000000      76.000000    1993.000000   
75%    1.375140e+06     111.000000     110.000000    2539.000000   
max    1.499999e+06     145.000000     186.000000    2956.000000   

       checkout_price     base_price  emailer_for_promotion  \
count   456548.000000  456548.000000          456548.000000   
mean       332.238933     354.156627               0.081152   
std        152.939723     160.715914               0.273069   
min          2.970000      55.350000               0.000000   
25%        228.950000     243.500000               0.000000   
50%        296.820000     310.460000               0.000000   
75%        445.230000     458.870000               0.000000   
max        866.270000     866.270000               1.000000   

       homepage_featured     num_orders      city_code    region_code  \
count       456548.00000  456548.000000  456548.000000  456548.000000   
mean             0.10920     261.872760     601.553399      56.614566   
std              0.31189     395.922798      66.195914      17.641306   
min              0.00000      13.000000     456.000000      23.000000   
25%              0.00000      54.000000     553.000000      34.000000   
50%              0.00000     136.000000     596.000000      56.000000   
75%              0.00000     324.000000     651.000000      77.000000   
max              1.00000   24299.000000     713.000000      93.000000   

             op_area  
count  456548.000000  
mean        4.083590  
std         1.091686  
min         0.900000  
25%         3.600000  
50%         4.000000  
75%         4.500000  
max         7.000000

We want to predict the number of orders for each meal at each center. So, the target variable is number of orders (num_orders).

### Check for missing data

In [201]:
# check if there is any missing values
food_data.isna().sum()

id                       0
week                     0
center_id                0
meal_id                  0
checkout_price           0
base_price               0
emailer_for_promotion    0
homepage_featured        0
num_orders               0
city_code                0
region_code              0
center_type              0
op_area                  0
meal_category            0
cuisine                  0
dtype: int64

There is no missing data in the dataset.

#### Features unique values

In [203]:
#Estimate number of unique values for some features
meals=food_data.meal_id.unique().size
centers=food_data.center_id.unique().size
city=food_data.city_code.unique().size
region=food_data.region_code.unique().size
centerType=food_data.center_type.unique().size
meal_cat=food_data.meal_category.unique().size
cuisine=food_data.cuisine.unique().size

print(f'center={centers}, meal={meals}, city= {city}, region= {region}, center_Type= {centerType}, meal_cat= {meal_cat}, cuisine= {cuisine}')

center=77, meal=51, city= 51, region= 8, center_Type= 3, meal_cat= 14, cuisine= 4


In [143]:
food_data.emailer_for_promotion.value_counts()

emailer_for_promotion
0    419498
1     37050
Name: count, dtype: int64

In [144]:
food_data.homepage_featured.value_counts()

homepage_featured
0    406693
1     49855
Name: count, dtype: int64

#### Checking the checkout price and base price to explore their relationship


In [145]:
checkout=len(food_data[food_data['checkout_price']>food_data['base_price']])
checkout

116101

In [146]:
base=len(food_data[food_data['checkout_price']<food_data['base_price']])
base

229687

In [147]:
check_base=len(food_data[food_data['checkout_price']==food_data['base_price']])
check_base

110760

The above calculation suggests that the checkout price varies in relation to the base price, with a substantial number of cases where the checkout price is either greater than or less than the base price. The almost equal count for cases where the checkout price equals the base price (110,760) further supports the idea that there isn't a specific consistent relationship between the checkout and base prices.

#### Checking the email promotion and page featured and thier relationship

In [149]:
    (food_data[food_data.emailer_for_promotion!= food_data.homepage_featured]
 .groupby('emailer_for_promotion')['homepage_featured']
 .value_counts())

emailer_for_promotion  homepage_featured
0                      1                    30624
1                      0                    17819
Name: count, dtype: int64

In [150]:
    (food_data[food_data.emailer_for_promotion== food_data.homepage_featured]
 .groupby('emailer_for_promotion')['homepage_featured']
 .value_counts())

emailer_for_promotion  homepage_featured
0                      0                    388874
1                      1                     19231
Name: count, dtype: int64

There seems to be a correlation between email promotions and homepage features. When they are not featured together, there are fewer instances. However, when they are featured together or not featured together, the counts are higher, indicating a potential relationship or alignment between these two features.